### PR TITLE
Your wish is my command...

### PR DESCRIPTION
**Pull request by NickLaMuro**
@carbonin As you requested :smile:

> This bit of code to translate the credentials is now in 4 places 😭 if any of you want to help me work out a way to DRY that up I'd be grateful.
>
> \- @carbonin (July 15, 2019)

DRYs up code for translating credentials.  Makes it as flexible as possible to fit into the multiple places it is being used.

Note:  commit is `[WIP]` until I can run specs for the previous areas of the code to confirm this at least works a little bit.

Links
-----

* Built on top of https://github.com/ManageIQ/manageiq/pull/18983

### CODE REVIEW

File: app/models/mixins/ansible_runner_auth_translations.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module AnsibleRunnerAuthTranslations
+  AUTH_TYPES = %i[credential vault_credential cloud_credential network_credential]
+
+  module_function
+
+  # Translates options hash with credential_ids into auth
+  #
+  # @param [Hash] options Hash containing the credential_ids
+  # @param [Hash] other_hash (options) Data to be updated if differs from <options>
```
Comment (NickLaMuro) [inline app/models/mixins/ansible_runner_auth_translations.rb:9]: > `# @param [Hash] other_hash ...`

#namingIsHard™

File: app/models/mixins/ansible_runner_auth_translations.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  # @param [Boolean] mutate_options (false) If passed in option keys should be removed, set to true
+  #
+  def translate_credentials!(options, other_hash: nil, mutate_options: false)
+    other_hash ||= options
+
+    AUTH_TYPES.each do |credential|
+      credential_sym         = "#{credential}_id".to_sym
+      credential_id          = mutate_options ? options.delete(credential_sym) : options[credential_sym]
+      other_hash[credential] = Authentication.find(credential_id).native_ref if credential_id.present?
```
Comment (NickLaMuro) [inline app/models/mixins/ansible_runner_auth_translations.rb:18]: I think I covered all of the corner cases with this by using `:other_hash` and `:mutate_options`, but please check my work.  I already forgot to assign this to `other_hash` instead of `options`.

Probably want to test this mixin directly and heavily just in case someone comes and changes this method and does something unexpected.

File: app/models/mixins/ansible_runner_auth_translations.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+module AnsibleRunnerAuthTranslations
+  AUTH_TYPES = %i[credential vault_credential cloud_credential network_credential]
+
+  module_function
+
+  # Translates options hash with credential_ids into auth
```
Comment (NickLaMuro) [inline app/models/mixins/ansible_runner_auth_translations.rb:6]: Along the same lines with my comment about adding specs, the docs on this probably could be fleshed out a bit too.

File: app/models/service_template_ansible_playbook.rb [hunk 4]
```diff
@@ -2,6 +2,8 @@ class ServiceTemplateAnsiblePlaybook < ServiceTemplateGeneric
   before_destroy :check_retirement_potential, :prepend => true
   around_destroy :around_destroy_callback, :prepend => true
 
+  include AnsibleRunnerAuthTranslations
```
Comment (NickLaMuro) [inline app/models/service_template_ansible_playbook.rb]: Fudged this one up... this needs to be in the class method space, so it should have been a `extend`, not `include`.  Fixing...

File: app/models/mixins/ansible_runner_auth_translations.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+module AnsibleRunnerAuthTranslations
+  AUTH_TYPES = %i[credential vault_credential cloud_credential network_credential]
+
+  module_function
```
Comment (NickLaMuro) [inline app/models/mixins/ansible_runner_auth_translations.rb:4]: By making this method a `module_function`, this can be called in two ways:

- By `include`ing it, and calling it as we have done currently (as if it were a method in the class)
- Calling it from it's `module` directly as if it was a class method (similar to `FileUtils`)

The second form would look like:

```ruby
AnsibleRunnerAuthTranslations.translate_credentials(options)
```

Which has the advantage of not adding methods to the classes we are writing this for, though, when adding a module that has `module_function`s, it should add them into the private space (I think...).

**Review by NickLaMuro — COMMENTED** [review]
Just providing some preliminary explanations about the changes.

**Comment by carbonin** [comment]
So I was thinking more along the lines of someone from GM's team helping to work out how we can dedup the whole processing in this part of the code base.

I don't really want to create a top-level mixin for one method that, realistically, I would like to get rid of entirely in time (there's no reason the change the keys since we're not passing them to tower anymore).

**Comment by NickLaMuro** [comment]
Fair enough.  My bad for misinterpreting.

Feel free to close then, since I think your broader plan makes more sense long term.
